File: engine/src/input.cc
#include "input.hh"

#include "SDL2/SDL.h"

namespace Hobby::Input {

static SDL_KeyCode ConvertToSDL(KeyType key);

bool IsKeyDown(KeyType key) {
  const uint8_t* state = SDL_GetKeyboardState(nullptr);
  SDL_Scancode scancode = SDL_GetScancodeFromKey(ConvertToSDL(key));
  return state[scancode] ? true : false;
}

static SDL_KeyCode ConvertToSDL(KeyType key) {
  switch (key) {
    case KeyType::UNKNOWN: return SDLK_UNKNOWN;
    case KeyType::F1: return SDLK_F1;
    case KeyType::F2: return SDLK_F2;
    case KeyType::F3: return SDLK_F3;
    case KeyType::F4: return SDLK_F4;
    case KeyType::F5: return SDLK_F5;
    case KeyType::F6: return SDLK_F6;
    case KeyType::F7: return SDLK_F7;
    case KeyType::F8: return SDLK_F8;
    case KeyType::F9: return SDLK_F9;
    case KeyType::F10: return SDLK_F10;
    case KeyType::F11: return SDLK_F11;
    case KeyType::F12: return SDLK_F12;
    case KeyType::F13: return SDLK_F13;
    case KeyType::F14: return SDLK_F14;
    case KeyType::F15: return SDLK_F15;
    case KeyType::F16: return SDLK_F16;
    case KeyType::F17: return SDLK_F17;
    case KeyType::F18: return SDLK_F18;
    case KeyType::F19: return SDLK_F19;
    case KeyType::F20: return SDLK_F20;
    case KeyType::F21: return SDLK_F21;
    case KeyType::F22: return SDLK_F22;
    case KeyType::F23: return SDLK_F23;
    case KeyType::F24: return SDLK_F24;
    case KeyType::ESCAPE: return SDLK_ESCAPE;
    case KeyType::BACKSPACE: return SDLK_BACKSPACE;
    case KeyType::TAB: return SDLK_TAB;
    case KeyType::CAPSLOCK: return SDLK_CAPSLOCK;
    case KeyType::RETURN: return SDLK_RETURN;
    case KeyType::UP: return SDLK_UP;
    case KeyType::DOWN: return SDLK_DOWN;
    case KeyType::LEFT: return SDLK_LEFT;
    case KeyType::RIGHT: return SDLK_RIGHT;
    case KeyType::PRINTSCREEN: return SDLK_PRINTSCREEN;
    case KeyType::SCROLLLOCK: return SDLK_SCROLLLOCK;
    case KeyType::INSERT: return SDLK_INSERT;
    case KeyType::HOME: return SDLK_HOME;
    case KeyType::PAGEUP: return SDLK_PAGEUP;
    case KeyType::PAGEDOWN: return SDLK_PAGEDOWN;
    case KeyType::END: return SDLK_END;
    case KeyType::LGUI: return SDLK_LGUI;
    case KeyType::RGUI: return SDLK_RGUI;
    case KeyType::LSHIFT: return SDLK_LSHIFT;
    case KeyType::RSHIFT: return SDLK_RSHIFT;
    case KeyType::LCTRL: return SDLK_LCTRL;
    case KeyType::RCTRL: return SDLK_RCTRL;
    case KeyType::LALT: return SDLK_LALT;
    case KeyType::RALT: return SDLK_RALT;
    case KeyType::KP_ENTER: return SDLK_KP_ENTER;
    case KeyType::DELETE: return SDLK_DELETE;
    case KeyType::KP_0: return SDLK_KP_0;
    case KeyType::KP_1: return SDLK_KP_1;
    case KeyType::KP_2: return SDLK_KP_2;
    case KeyType::KP_3: return SDLK_KP_3;
    case KeyType::KP_4: return SDLK_KP_4;
    case KeyType::KP_5: return SDLK_KP_5;
    case KeyType::KP_6: return SDLK_KP_6;
    case KeyType::KP_7: return SDLK_KP_7;
    case KeyType::KP_8: return SDLK_KP_8;
    case KeyType::KP_9: return SDLK_KP_9;
    case KeyType::KP_DIVIDE: return SDLK_KP_DIVIDE;
    case KeyType::KP_MULTIPLY: return SDLK_KP_MULTIPLY;
    case KeyType::KP_MINUS: return SDLK_KP_MINUS;
    case KeyType::KP_PLUS: return SDLK_KP_PLUS;
    case KeyType::K0: return SDLK_0;
    case KeyType::K1: return SDLK_1;
    case KeyType::K2: return SDLK_2;
    case KeyType::K3: return SDLK_3;
    case KeyType::K4: return SDLK_4;
    case KeyType::K5: return SDLK_5;
    case KeyType::K6: return SDLK_6;
    case KeyType::K7: return SDLK_7;
    case KeyType::K8: return SDLK_8;
    case KeyType::K9: return SDLK_9;
    case KeyType::A: return SDLK_a;
    case KeyType::B: return SDLK_b;
    case KeyType::C: return SDLK_c;
    case KeyType::D: return SDLK_d;
    case KeyType::E: return SDLK_e;
    case KeyType::F: return SDLK_f;
    case KeyType::G: return SDLK_g;
    case KeyType::H: return SDLK_h;
    case KeyType::I: return SDLK_i;
    case KeyType::J: return SDLK_j;
    case KeyType::K: return SDLK_k;
    case KeyType::L: return SDLK_l;
    case KeyType::M: return SDLK_m;
    case KeyType::N: return SDLK_n;
    case KeyType::O: return SDLK_o;
    case KeyType::P: return SDLK_p;
    case KeyType::Q: return SDLK_q;
    case KeyType::R: return SDLK_r;
    case KeyType::S: return SDLK_s;
    case KeyType::T: return SDLK_t;
    case KeyType::U: return SDLK_u;
    case KeyType::V: return SDLK_v;
    case KeyType::W: return SDLK_w;
    case KeyType::X: return SDLK_x;
    case KeyType::Y: return SDLK_y;
    case KeyType::Z: return SDLK_z;
    case KeyType::SPACE: return SDLK_SPACE;
    case KeyType::COMMA: return SDLK_COMMA;
    case KeyType::PERIOD: return SDLK_PERIOD;
    case KeyType::SLASH: return SDLK_SLASH;
    case KeyType::SEMICOLON: return SDLK_SEMICOLON;
    case KeyType::QUOTE: return SDLK_QUOTE;
    case KeyType::LEFTBRACKET: return SDLK_LEFTBRACKET;
    case KeyType::RIGHTBRACKET: return SDLK_RIGHTBRACKET;
    case KeyType::BACKSLASH: return SDLK_BACKSLASH;
    case KeyType::BACKQUOTE: return SDLK_BACKQUOTE;
    case KeyType::MINUS: return SDLK_MINUS;
    case KeyType::EQUALS: return SDLK_EQUALS;
    default: break;
  }

  return SDLK_UNKNOWN;
}

} // namespace Hobby::Input

File: src/lua/input.lua
function hobby.input.getvector(left, right, up, down)
  hobby.matchtype(left, "number")
  hobby.matchtype(right, "number")
  hobby.matchtype(up, "number")
  hobby.matchtype(down, "number")

  local ix, iy = 0, 0
  if hobby.input.iskeydown(left) then ix = ix - 1 end
  if hobby.input.iskeydown(right) then ix = ix + 1 end
  if hobby.input.iskeydown(up) then iy = iy - 1 end
  if hobby.input.iskeydown(down) then iy = iy + 1 end
  return hobby.vec.normalize(ix, iy)
end

hobby.keyevent = hobby.event()
hobby.keypressed = hobby.event()
hobby.keyreleased = hobby.event()

function hobby.onkeypressed(key, isRepeat)
  hobby.keypressed:call(key, isRepeat)
  hobby.keyevent:call(true, key, isRepeat)
end

function hobby.onkeyreleased(key)
  hobby.keyreleased:call(key)
  hobby.keyevent:call(false, key, nil)
end

local keynames = {}

function hobby.input.keytostring(key)
  local name = keynames[key]
  if not name then
    error("Invalid key.", 1)
  end

  return name
end

local keys = {
  Space        = 32,
  Apostrophe   = 39,
  Comma        = 44,
  Minus        = 45,
  Period       = 46,
  Slash        = 47,
  K0           = 48,
  K1           = 49,
  K2           = 50,
  K3           = 51,
  K4           = 52,
  K5           = 53,
  K6           = 54,
  K7           = 55,
  K8           = 56,
  K9           = 57,
  Semicolon    = 59,
  Equal        = 61,
  A            = 65,
  B            = 66,
  C            = 67,
  D            = 68,
  E            = 69,
  F            = 70,
  G            = 71,
  H            = 72,
  I            = 73,
  J            = 74,
  K            = 75,
  L            = 76,
  M            = 77,
  N            = 78,
  O            = 79,
  P            = 80,
  Q            = 81,
  R            = 82,
  S            = 83,
  T            = 84,
  U            = 85,
  V            = 86,
  W            = 87,
  X            = 88,
  Y            = 89,
  Z            = 90,
  LeftBracket  = 91,
  Backslash    = 92,
  RightBracket = 93,
  GraveAccent  = 96,
  World1       = 161,
  World2       = 162,
  Escape       = 256,
  Enter        = 257,
  Tab          = 258,
  Backspace    = 259,
  Insert       = 260,
  Delete       = 261,
  Right        = 262,
  Left         = 263,
  Down         = 264,
  Up           = 265,
  PageUp       = 266,
  PageDown     = 267,
  Home         = 268,
  End          = 269,
  CapsLock     = 280,
  ScrollLock   = 281,
  NumLock      = 282,
  PrintScreen  = 283,
  Pause        = 284,
  F1           = 290,
  F2           = 291,
  F3           = 292,
  F4           = 293,
  F5           = 294,
  F6           = 295,
  F7           = 296,
  F8           = 297,
  F9           = 298,
  F10          = 299,
  F11          = 300,
  F12          = 301,
  F13          = 302,
  F14          = 303,
  F15          = 304,
  F16          = 305,
  F17          = 306,
  F18          = 307,
  F19          = 308,
  F20          = 309,
  F21          = 310,
  F22          = 311,
  F23          = 312,
  F24          = 313,
  F25          = 314,
  Kp0          = 320,
  Kp1          = 321,
  Kp2          = 322,
  Kp3          = 323,
  Kp4          = 324,
  Kp5          = 325,
  Kp6          = 326,
  Kp7          = 327,
  Kp8          = 328,
  Kp9          = 329,
  KpDecimal    = 330,
  KpDivide     = 331,
  KpMultiply   = 332,
  KpSubtract   = 333,
  KpAdd        = 334,
  KpEnter      = 335,
  KpEqual      = 336,
  LeftShift    = 340,
  LeftControl  = 341,
  LeftAlt      = 342,
  LeftSuper    = 343,
  RightShift   = 344,
  RightControl = 345,
  RightAlt     = 346,
  RightSuper   = 347,
  Menu         = 348,
}

hobby.LMB = 1
hobby.RMB = 2
hobby.MMB = 3
hobby.XB1 = 4
hobby.XB2 = 5

for k, v in pairs(keys) do
  hobby[k] = v
  keynames[v] = k
end
